docs/Configuration.md
# Configuration

## Modules
The configuration of each module is described in its respective sub-configuration file:
* [Inputs.md](Inputs.md)
* [Outputs.md](Outputs.md)
* [Sources.md](Sources.md)
* [Hooks.md](Hooks.md)

## Privileged mode

To start privileged mode before pulling the configuration, Oxidized needs to
send the enable command. You can globally enable this, by adding the following
snippet to the global section of the configuration file.

```yaml
vars:
   enable: S3cre7
```

## Removing secrets

To strip out secrets from configurations before storing them, Oxidized needs the
`remove_secret` flag. You can globally enable this by adding the following
snippet to the global section of the configuration file.

```yaml
vars:
  remove_secret: true
```

Device models that contain substitution filters to remove sensitive data will
now be run on any fetched configuration.

As a partial example from ios.rb:

```ruby
  cmd :secret do |cfg|
    cfg.gsub! /^(snmp-server community).*/, '\\1 <configuration removed>'
    # ...
    cfg
  end
```

The above strips out snmp community strings from your saved configs.

**NOTE:** Removing secrets reduces the usefulness as a full configuration backup, but it may make sharing configs easier.

## Timeout and Time limit
You can configure when oxidized will `timeout` while fetching a configuration
(default: 20 seconds), and how much absolute time (`timelimit`) the fetching
is allowed to last (default: 300 seconds, or 5 minutes):

* `timeout`: Maximum time to wait for a single operation during config fetching.
  Not every input module has an implemented timeout.
* `timelimit`: Maximum total time allowed for the entire fetch job. It is
  independent of input modules and will always be enforced.

If `timelimit`is reached, the fetch job will be killed and will produce a
warning. The job status will be set to `timelimit`.

```yaml
timeout: 20
timelimit: 300
```

## Advanced Configuration

Below is an advanced example configuration.

You will be able to (optionally) override options per device.
The router.db format used is `hostname:model:username:password:enable_password`.
Hostname and model will be the only required options, all others override the
global configuration sections.

Custom model names can be mapped to an oxidized model name with a string or
a regular expression.


```yaml
---
username: oxidized
password: S3cr3tx
model: junos
interval: 3600 #interval in seconds, when 0 is configured no fetch config is done at initial start and after
debug: false
threads: 30 # maximum number of threads
# use_max_threads:
# false - the number of threads is selected automatically based on the interval option, but not more than the maximum
# true - always use the maximum number of threads
use_max_threads: false
timeout: 20
timelimit: 300
retries: 3
prompt: !ruby/regexp /^([\w.@-]+[#>]\s?)$/
crash:
  directory: ~/.config/oxidized/crashes
  hostnames: false
vars:
  enable: S3cr3tx
groups: {}
extensions:
  oxidized-web:
    load: true
    # Bind to any IPv4 interface
    listen: 0.0.0.0
    # Bind to port 8888 (default)
    port: 8888
    # Prefix prod to the URL, so http://oxidized.full.domain/prod/
    url_prefix: prod
    # virtual hosts to listen to (others will be denied)
    vhosts:
      - localhost
      - 127.0.0.1
      - oxidized
      - oxidized.full.domain
pid: ~/.config/oxidized/oxidized.pid
input:
  default: ssh, telnet
  debug: false
  ssh:
    secure: false
output:
  default: git
  git:
      user: Oxidized
      email: oxidized@example.com
      repo: "~/.config/oxidized/oxidized.git"
source:
  default: csv
  csv:
    file: ~/.config/oxidized/router.db
    delimiter: !ruby/regexp /:/
    map:
      name: 0
      model: 1
      username: 2
      password: 3
    vars_map:
      enable: 4
model_map:
  cisco: ios
  juniper: junos
  !ruby/regexp /procurve/: procurve
logger:
  # The default level is :info
  # level: :info
  appenders:
    - type: syslog
      level: :error
    - type: stdout
      level: :warn
    - type: file
      file: ~/.config/oxidized/info.log
```

## Advanced Group Configuration

For group specific credentials

```yaml
groups:
  mikrotik:
    username: admin
    password: blank
  ubiquiti:
    username: ubnt
    password: ubnt
```

Model specific variables/credentials within groups

```yaml
groups:
  foo:
    models:
      arista:
        username: admin
        password: password
        vars:
          ssh_keys: "~/.ssh/id_rsa_foo_arista"
      vyatta:
        vars:
          ssh_keys: "~/.ssh/id_rsa_foo_vyatta"
  bar:
    models:
      routeros:
        vars:
          ssh_keys: "~/.ssh/id_rsa_bar_routeros"
      vyatta:
        username: admin
        password: pass
        vars:
          ssh_keys: "~/.ssh/id_rsa_bar_vyatta"
```

For mapping multiple group values to a common name, you can use strings and
regular expressions:

```yaml
group_map:
  alias1: groupA
  alias2: groupA
  alias3: groupB
  alias4: groupB
  !ruby/regexp /specialgroup/: groupS
  aliasN: groupZ
  # ...
```

add group mapping to a source

```yaml
source:
  # ...
  <source>:
    # ...
    map:
      model: 0
      name: 1
      group: 2
```

For model specific credentials

You can add 'username: nil' if the device only expects a Password at prompt.

```yaml
models:
  junos:
    username: admin
    password: password
  ironware:
    username: admin
    password: password
    vars:
      enable: enablepassword
  apc_aos:
    username: apc
    password: password
  cisco:
    username: nil
    password: pass
```

## Options (credentials, vars, etc.) precedence:
From least to most important:
- global options
- model specific options
- group specific options
- model specific options in groups
- options defined on single nodes

More important options overwrite less important ones if they are set.

## oxidized-web: RESTful API and web interface

The RESTful API and web interface are enabled by installing the `oxidized-web`
gem and configuring the `extensions.oxidized-web:` section in the configuration
file. You can set the following parameter:
- `load`: `true`/`false`: Enables or disables the `oxidized-web` extension
  (default: `false`)
- `listen`: Specifies the interface to bind to (default: `127.0.0.1`). Valid
  options:
  - `127.0.0.1`: Allows IPv4 connections from localhost only
  - `'[::1]'`: Allows IPv6 connections from localhost only
  - `<IPv4-Address>` or `'[<IPv6-Address>]'`: Binds to a specific interface
  - `0.0.0.0`: Binds to any IPv4 interface
  - `'[::]'`:  Binds to any IPv4 and IPv6 interface
- `port`: Specifies the TCP port to listen to (default: `8888`)
- `url_prefix`: Defines a URL prefix (default: no prefix)
- `vhosts`: A list of virtual hosts to listen to. If not specified, it will
  respond to any virtual host.

> [!NOTE]
> The old syntax `rest: 127.0.0.1:8888/prefix` is still supported but
> deprecated. It produces a warning and won't be suported in future releases.
>
> If the `rest` configuration is used, the extensions.oxidized-web will be
> ignored.

> [!NOTE]
> You need oxidized-web version 0.16.0 or later to use the
> extentions.oxidized-web configuration


```yaml
# Listen on http://[::1]:8888/
extensions:
  oxidized-web:
    load: true
    listen: '[::1]'
    port: 8888
```

```yaml
# Listen on http://127.0.0.1:8888/
extensions:
  oxidized-web:
    load: true
    listen: 127.0.0.1
    port: 8888
```

```yaml
# Listen on http://[2001:db8:0:face:b001:0:dead:beaf]:8888/oxidized/
extensions:
  oxidized-web:
    load: true
    listen: '[2001:db8:0:face:b001:0:dead:beaf]'
    port: 8888
    url_prefix: oxidized
```

```yaml
# Listen on http://10.0.0.1:8000/oxidized/
extensions:
  oxidized-web:
    load: true
    listen: 10.0.0.1
    port: 8000
    url_prefix: oxidized
```

```yaml
# Listen on any interface to http://oxidized.rocks:8888 and
# http://oxidized:8888
extensions:
  oxidized-web:
    load: true
    listen: '[::]'
    url_prefix: oxidized
    vhosts:
     - oxidized.rocks
     - oxidized
```

## Triggered backups

A node can be moved to head-of-queue via the REST API `GET/PUT
/node/next/[NODE]`. This can be useful to immediately schedule a fetch of the
configuration after some other event such as a syslog message indicating a
configuration update on the device.

In the default configuration this node will be processed when the next job
worker becomes available, it could take some time if existing backups are in
progress. To execute moved jobs immediately a new job can be added
automatically:

```yaml
next_adds_job: true
```

This will allow for a more timely fetch of the device configuration.

## Disabling DNS resolution

In some instances it might not be desirable to attempt to resolve names of
nodes. One such use case is when nodes are accessed through an SSH proxy, where
the remote end resolves the names differently than the host on which Oxidized
runs would.

Names can instead be passed verbatim to the input:

```yaml
resolve_dns: false
```

## Environment variables

You can use some environment variables to change default root directories values.

* `OXIDIZED_HOME` may be used to set oxidized configuration directory, which defaults to `~/.config/oxidized`
* `OXIDIZED_LOGS` may be used to set oxidzied logs and crash directories root, which default to `~/.config/oxidized`

## Logging
Oxidized supports parallel logging to different systems (appenders). The
following appenders are currently supported:
- `stderr`: log to standard error (this is the default)
- `stdout`: log to standard output
- `file`: log to a file
- `syslog`: log to syslog

> `stderr` and `stdout` are mutually exclusive and will produce a warning if used
> simultaneously.

> You can configure as many file appenders as you wish.

You can set a log level globally and/or for each appender.
- The global log level will limit which log messages are accepted, depending
  on their level.
  - The default global log level is `:info`.
  - If you set `debug: true` in the configuration, the global log level will be
    forced to `:debug`.
- The appender log level limits which log messages are displayed by the
  appender, depending on their level.
  - The default is `:trace`.


> Available log levels: `:trace`, `:debug`, `:info`, `:warn`,
> `:error` and `:fatal`

Here is a configuration example logging `:error` to syslog, `:warn` to stdout
and `:info` to `~/.config/oxidized/info.log`:

```yaml
logger:
  # Default level
  # level: :info
  appenders:
    - type: syslog
      level: :error
    - type: stdout
      level: :warn
    - type: file
      # Default level is :trace, so we get the logs in the default level (:info)
      file: ~/.config/oxidized/info.log
```

If you want to log :trace to a file and `:info` to stdout, you must set the
global log level to `:trace`, and limit the stdout appender to `:info`:

```yaml
logger:
  level: :trace
  appenders:
    - type: stdout
      level: :info
    - type: file
      file: ~/.config/oxidized/trace.log
```

### Change log level
You can change the global log level of oxidized by sending a SIGUSR2 to
the process:
```
kill -SIGUSR2 424242
```
It will rotate between the log levels and log a warning with the new level
(you won't see the warning when the log level is `:fatal` or `:error`):
```
2025-06-30 15:25:27.972881 W [109750:2640] SemanticLogger -- Changed global default log level to :warn
```

If you specified a log level for an appender, this log level won't be
changed.

> :warning: **Warning** You need oxidized-web 0.17.0 and above for this or
> it will kill the whole oxidized application.

### Dump running threads
With the SIGTTIN signal, oxidized will log a backtrace for each of its threads.
```
kill -SIGTTIN 424242
```

The threads used to fetch the configs are named `Oxidized::Job 'hostname'`:

```
2025-06-30 15:32:22.293047 W [110549:2640 core.rb:76] Thread Dump -- Backtrace:
/home/xxx/oxidized/lib/oxidized/core.rb:76:in `sleep'
/home/xxx/oxidized/lib/oxidized/core.rb:76:in `block in run'
(...)
2025-06-30 15:32:22.293409 W [110549:Oxidized::Job 'host2' ssh.rb:127] Thread Dump -- Backtrace:
/home/xxx/oxidized/lib/oxidized/input/ssh.rb:127:in `sleep'
/home/xxx/oxidized/lib/oxidized/input/ssh.rb:127:in `block (2 levels) in expect'
```

## Metadata
You can include some metadata in your model outputs, for this you have to set
the variable `metadata` to `true`:
```yaml
vars:
  metadata: true
```

As every [variable](#options-credentials-vars-etc-precedence), you can set it on
model, group and even node level.

By default this will produce
`"%{comment}Fetched by Oxidized with model %{model} from host %{name} [%{ip}]\n"`
at the first line of every model output. Some models with specific needs (XML
for example) will save the metadata differently (for example, OpnSense and
PfSense save an XML comment at the end of the model).

### Customize metadata
You can customize the metadata produced by setting the varibles `metadata_top`
(top of the file) and `metadata_bottom` (bottom of the file).

These variables accept string templates, and you can include newline characters
(\n) to control formatting.

Both `metadata_top` and `metadata_bottom` support interpolation of dynamic values
using the following substitution templates:
  - `%{model}`: name of the Oxidized model
  - `%{name}`: name of the node
  - `%{ip}`: IP address of the node
  - `%{group}`: group name of the node
  - `%{comment}`: comment string used in the model output (`# `)
  - `%{year}`: current year (`2025`)
  - `%{month}`: current month, zero-padded (`03` for March)
  - `%{day}`: current day, zero-padded (`09`)
  - `%{hour}`: current hour (24-hour format, zero-padded)
  - `%{minute}`: current minute, zero-padded
  - `%{second}`: current second, zero-padded

Example:
```yaml
vars:
  metadata: true
  metadata_top: "%{comment}Model: %{model}; Device %{name} [%{ip}] at %{year}-%{month}-%{day} %{hour}:%{minute}:%{second}\n"
```

### Customize metadata in models
When writing a custom metadata for a model, you can default to
`vars("metadata_*")` or the model default. You need to interpolate the strings
with interpolate_string. This example is taken from OpnSense, and makes an 
XML comment of the default strings, with precedence for vars("metadata_bottom"),
as the XML comment is situated at the bottom.

```ruby
  metadata :bottom do
    xmlcomment interpolate_string(
      vars("metadata_bottom") ||
      vars("metadata_top") ||
      Oxidized::Model::METADATA_DEFAULT
    )
  end
```

You can also change the metadata in the models in your configuration directory with
[monkey patching](Creating-Models.md#monkey-patching-blocks-in-existing-models).
This can be done in two ways:

1. Interpolation string:
```ruby
require 'oxidized/model/ios.rb'

class IOS
  metadata :top, "%{comment}Model: %{model}\n"
  metadata :bottom, "%{comment}Will be placed at the end of the output\n"
end
```

2. For advanced code, you can use a code block, and access every variable of
   the model:

```ruby
require 'oxidized/model/opnsense.rb'
class OpnSense
  metadata :top do
    xmlcomment "Model:#{self.class.name}, hostname: #{@node.name}, ip:#{@node.ip}"
  end

  metadata :bottom, nil
end
```

Remove a previous metadata by setting it to `nil`.

## Store configuration only on significant changes
Some devices produce configuration changes even though nothing relevant
changed. For example, Cisco IOS produces a `Last configuration change at` as
soon as you exit config mode, and FortiOS encrypts its passwords with a
different salt on every run.

By setting the [variable](#options-credentials-vars-etc-precedence)
`output_store_mode` to `on_significant`, you can tell Oxidized only to
store the configuration when significant changes occurred. The default is to
always store the configuration.
```yaml
vars:
  output_store_mode: on_significant
```

For this to work, the model must implement `cmd :significant_changes`:
```ruby
  cmd :significant_changes do |cfg|
    cfg.reject_lines [
      'Last configuration change at',
      'NVRAM config last updated at'
    ]
  end
```

Note that store on significant change only applies to the main configuration,
and will not affect
[output types](Creating-Models.md#advanced-feature-output-type)


spec/cli/data/exhaustive_config#output.txt
```yaml
---
username: oxidized
password: [REDACTED]
model: junos
resolve_dns: true
interval: 3600
debug: true
run_once: false
threads: 10
use_max_threads: true
timeout: 20
timelimit: 300
retries: 3
prompt: !ruby/regexp /^([\w.@-]+[#>]\s?)$/
next_adds_job: false
vars:
  metadata: true
  metadata_top: "%{comment}Model: %{model}; Device %{name} [%{ip}] at %{year}-%{month}-%{day}
    %{hour}:%{minute}:%{second}\n"
  output_store_mode: on_significant
groups: {}
group_map: {}
models:
  openwrt:
    username: root
    password: [REDACTED]
  routeros:
    username: admin
    password: [REDACTED]
  asa:
    vars:
      enable: [REDACTED]
  fortios:
    vars:
      fullconfig: true
pid: "/home/cheramr/.config/oxidized/pid"
extensions:
  oxidized-web:
    load: true
    hide_node_vars:
    - enable
    - password
    - username
crash:
  directory: "/home/cheramr/.config/oxidized/crashes"
  hostnames: false
stats:
  history_size: 10
input:
  default: ssh,scp
  debug: false
  ssh:
    secure: false
  ftp:
    passive: true
  utf8_encoded: true
output:
  default: git
  git:
    user: Oxidized
    email: oxidized@example.com
    repo: "~/.config/oxidized/oxidized.git"
    single_repo: true
  file:
    directory: "/home/cheramr/.config/oxidized/configs"
source:
  default: csv
  csv:
    file: "~/.config/oxidized/devicelist.db"
    delimiter: !ruby/regexp /:/
    map:
      name: 0
      model: 1
      ip: 2
      group: 3
    vars_map:
      hullo: 4
      enable: [REDACTED]
    gpg: false
model_map:
  juniper: junos
  cisco: ios
logger:
  appenders:
  - type: syslog
  - type: stdout
hooks:
  push_to_remote:
    type: githubrepo
    events:
    - post_store
    remote_repo: oxidized@localhost:oxidized.git
    privatekey: [REDACTED]
    publickey: "/home/cheramr/.ssh/id_ed25519.pub"
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).

## [0.37.0 - 2026-05-20]
### Added
- routeros: support store mode `on_significant` (@infabo)
- model for Grandstream HT8xx (@mklopocki)
- Add --support option to gather system diagnostics (@robertcheramy)

### Changed
- input/ssh: validate that cmd is a String. See #3700 (@robertcheramy)
- junos: changed "show system license" regex to reduce notification noise for PTX routers. See #3794 (@ctomkow)
- ios: keep motherboard information in show version. Closes #3798 (@robertcheramy)
- aoscx: use oxidized filtering instead of "show system | exclude...". (@robertcheramy)
- input/*: rewrite debug logging; remove unused logging; input/ssh: write a YAML simulation file (@robertcheramy)
- truenas: capture app, replication, cloudsync configurations without constant changes. See #3795 (@neilschelly)
- Set Rake::TestTask to `verbose = false` since the behavior changed with rake 13.4.2 (@robertcheramy)

### Fixed
- VyOS: detect community string in SNMP traps. Fixes: #3793 (@nicolasberens)
- fortigate: show Part No. of SFPs. Fixes: #3800 (@robertcheramy)
- aoscx: Hide total power consumption (stacking). Fixes #3801 (@robertcheramy)
- linuxgeneric: fix prompt, telnet username and clean ANSI Escape Codes (@robertcheramy)
- cumulus: fix prompt (make : and space optional). Fixes #3812 (@robertcheramy)
- dlinknexgen: NULL byte handling. Fixes #3816 (@ziotibia81)
- nxos: hide snmpv3 priv key when remove_secret is true. Fixes: #3805 (@robertcheramy)
- tpink: use `\r\n` when sending commands and set `terminal length 0`. Fixes #3804 (@robertcheramy)


## [0.36.0 - 2026-03-31]
### Release Notes
The fortios model has been split into fortigate and fortios. You need the new fortigate model for FortiGate firewalls. Be sure to check the [Fortinet model notes](docs/Model-Notes/Fortinet.md) before upgrading.

The SCP gem is now an optional dependency as it will rarely be used - you must install it if you need it. It is still included in the docker image.

We may rework models in the future to clean up duplicated code (by using the macros or by inheriting from the Defacto model). To make sure your favorite model doesn’t break, please share [simulation files](/docs/DeviceSimulation.md) via issues or, preferably, pull requests, so we can detect when a change breaks a model.

### Added
- Support https mode, headers, and ssl verify for HTTP output (@bahirul)
- String refinements: introduce `keep_lines` and `reject_lines` methods (@robertcheramy)
- Support for storing configurations only on significant changes (@robertcheramy)
- Add support for Ivanti Secure Connect ISA models (@candleflip)
- smartbyte: new model for SmartByte switches (@freddy36)
- Support multiple input (@robertcheramy)
- apcaos model with SSH + SCP capabilities, deprecates apc_aos (@robertcheramy)
- exalink model for specific Cisco ExaLink Fusion (formerly Exablaze) switches (@obol89)
- source_node_transform hook, allows user to manipulate node data when loading from source (@ytti)
- docker image: publish major version tag (e.g. `0`) in addition to full semver tag on releases (@infabo)
- introduce the defacto model and macros for models (@ytti)

### Changed
- Refactored models: Use `keep_lines` and `reject_lines` in aosw, arubainstant, asa, efos, firelinuxos, fsos, ironware, mlnxos and perle (@robertcheramy)
- Refactor SSH and SCP into a common class SSHBase. Fixes #3597 (@robertcheramy)
- Modified models to support store mode on significant changes: ios, fortios, perle, ndms (@robertcheramy, @furriest)
- fortios: model rewrite and split into fortios and fortigate. Fixes #3680 (@robertcheramy)
- fortigate: Add PSU & SFP inventory (@robertcheramy)
- fortigate: move var fortios_autoupdate (deprecated) to fortigate_autoupdate (@robertcheramy)
- netgear: extended login and pager detection to add support for GS728TPv2 and GS752TPv2 (@weberc)
- comware: Hide snmp secrets for Comware (@iriseden)
- Aruba-CX : Hide secrets for Aruba-CX (@iriseden)

### Fixed
- VyOS: Only remove SNMP community, not route-maps. Fixes #3735 (@systeembeheerder)
- apc_aos: set comment to "; " to match comments in config.ini (@robertcheramy)
- h3c: fix overly permissive prompt regexp causing false matches. Fixes #3673 (@robertcheramy)
- extra/device2yaml.rb: fix \r being removed at end of line (@robertcheramy)
- perle: remove trailing \r (the device sends \r\r\n) (@robertcheramy)
- Reintroduce support for Ruby 3.0. Fixes #3688 (@robertcheramy)
- githubrepo: fix authentication with ssh-agent not working. Fixes #3420 (@robertcheramy)
- fastiron: adjust prompt to account for stacks, remove time from stack output. Fixes #3106 (@ManoftheSea)
- interval: fix fetching device configuration at oxidized start when interval is 0. Fixes #3746 (@tgr229)
- voss: more cleanup of constantly changing values (fan & temp) for at least Extreme Networks 7520-48Y-8C-FabricEngine (8.10.5.0) (@irrwitzer42)
- truenas: Added retry logic to use sudo for reading/dumping the configuration database if needed. Fixes #3767 (@neilschelly)
- aoscx: update regex to include 'N/A' in FAN speed parsing (@solrac200, @robertcheramy)
- nxos: show inventory for older models. Fixes #3779 (@scamp)

## [0.35.0 - 2025-12-04]
### Release Notes
- VyOS now has it's own model and should be used for supported VyOS versions instead of the Vyatta model.
- AosCX has been reworked and may break old OS versions. Submit an issue along with a YAML Simulation File if you encounter problems.
- TiMOS (deprecated model) has been removed. Use SROS.
- FortiOs will be reworked in release 0.36 (Issue #3680). Subscribe to the issue if you want to be informed and test the model before the release.
- ~~Support for Ruby 3.1 will be discontinued in release 0.36 (Issue #3688) if no one objects.~~

### Added
- Allow setting timeout on per node basis. Closes #3612 (@ytti)

lib/oxidized/config.rb
module Oxidized
  require 'asetus'
  class NoConfig < OxidizedError; end
  class InvalidConfig < OxidizedError; end

  class Config
    ROOT       = ENV['OXIDIZED_HOME'] || File.join(Dir.home, '.config', 'oxidized')
    CRASH      = File.join(ENV['OXIDIZED_LOGS'] || ROOT, 'crash')
    LOG        = File.join(ENV['OXIDIZED_LOGS'] || ROOT, 'logs')
    INPUT_DIR  = File.join Directory, %w[lib oxidized input]
    OUTPUT_DIR = File.join Directory, %w[lib oxidized output]
    MODEL_DIR  = File.join Directory, %w[lib oxidized models]
    SOURCE_DIR = File.join Directory, %w[lib oxidized source]
    HOOK_DIR   = File.join Directory, %w[lib oxidized hook]
    SLEEP      = 1

    def self.load(cmd_opts = {})
      usrdir = File.expand_path(cmd_opts[:home_dir] || Oxidized::Config::ROOT)
      cfgfile = cmd_opts[:config_file] || 'config'
      # configuration file with full path as a class instance variable
      @configfile = File.join(usrdir, cfgfile)
      asetus = Asetus.new(name: 'oxidized', load: false, usrdir: usrdir, cfgfile: cfgfile)
      Oxidized.asetus = asetus

      asetus.default.username      = 'username'
      asetus.default.password      = 'password'
      asetus.default.model         = 'junos'
      asetus.default.resolve_dns   = true # if false, don't resolve DNS to IP
      asetus.default.interval      = 3600
      asetus.default.debug         = false
      asetus.default.run_once      = false
      asetus.default.threads       = 30
      asetus.default.use_max_threads = false
      asetus.default.timeout       = 20
      asetus.default.timelimit     = 300
      asetus.default.retries       = 3
      asetus.default.prompt        = /^([\w.@-]+[#>]\s?)$/
      asetus.default.next_adds_job = false            # if true, /next adds job, so device is fetched immmeiately
      asetus.default.vars          = {}               # could be 'enable'=>'enablePW'
      asetus.default.groups        = {}               # group level configuration
      asetus.default.group_map     = {}               # map aliases of groups to names
      asetus.default.models        = {}               # model level configuration
      asetus.default.pid           = File.join(Oxidized::Config::ROOT, 'pid')

      # Extentions
      asetus.default.extensions['oxidized-web'].load = false

      asetus.default.crash.directory = File.join(Oxidized::Config::ROOT, 'crashes')
      asetus.default.crash.hostnames = false

      asetus.default.stats.history_size = 10
      asetus.default.input.default      = 'ssh, telnet'
      asetus.default.input.debug        = false # or String for session log file
      asetus.default.input.ssh.secure   = false # complain about changed certs
      asetus.default.input.ftp.passive  = true  # ftp passive mode
      asetus.default.input.utf8_encoded = true  # configuration is utf8 encoded or ascii-8bit

      asetus.default.output.default = 'file'  # file, git
      asetus.default.source.default = 'csv'   # csv, sql

      asetus.default.model_map = {
        'juniper' => 'junos',
        'cisco'   => 'ios'
      }

      begin
        asetus.load # load system+user configs, merge to Config.cfg
      rescue StandardError => e
        raise InvalidConfig, "Error loading config: #{e.message}"
      end

      raise NoConfig, "edit #{@configfile}" if asetus.create

      # override if comand line flag given
      asetus.cfg.debug = cmd_opts[:debug] if cmd_opts[:debug]

      asetus
    end

    class << self
      attr_reader :configfile
    end
  end

  class << self
    attr_accessor :mgr, :hooks
  end
end
